File: src/main/java/pro/sky/java/service/LinkServiceImpl.java
package pro.sky.java.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.RandomStringUtils;
import org.springframework.stereotype.Service;
import pro.sky.java.domain.Link;
import pro.sky.java.repository.LinkRepository;
import reactor.core.publisher.Mono;

@Slf4j
@Service
@RequiredArgsConstructor
public class LinkServiceImpl implements LinkService {

    private final LinkRepository repository;

    @Override
    public Mono<Link> convertToLink(String original) {
        String shorten = createShortLink();
        return repository.save(shorten, original)
                .flatMap(v -> Mono.just(new Link(shorten, original)));
    }

    @Override
    public Mono<String> getOriginal(String shortUrl) {
        return repository.find(shortUrl);
    }

    private String createShortLink() {
        return RandomStringUtils.randomAlphanumeric(6);
    }
}

File: src/main/kotlin/com/github/disterru/linksshortener/service/LinkServiceImpl.kt
package com.github.disterru.linksshortener.service

import com.github.disterru.linksshortener.domain.Link
import com.github.disterru.linksshortener.repository.LinkRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.apache.commons.lang3.RandomStringUtils
import org.springframework.stereotype.Service

@Service
class LinkServiceImpl(
    private val repository: LinkRepository
) : LinkService {

    override suspend fun convertToLink(original: String) = withContext(Dispatchers.IO) {
        val shorten = createShortLink()
        repository.save(shorten, original)
        Link(original, shorten)
    }

    override suspend fun getOriginal(shortUrl: String) = withContext(Dispatchers.IO) {
        repository.find(shortUrl)
    }

    private fun createShortLink() = RandomStringUtils.randomAlphanumeric(6)
}